import sys
from PyQt5 import QtWidgets
from ui_main import Ui_MainWindow
from face_recognition import FaceRecognitionSystem


class MainApp(QtWidgets.QMainWindow):
    def __init__(self):
        super(MainApp, self).__init__()
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)

        self.recognizer = FaceRecognitionSystem()

        # –ü—Ä–∏–≤—è–∑–∫–∞ –∫–Ω–æ–ø–æ–∫
        self.ui.btnRegister.clicked.connect(self.register_user)
        self.ui.btnRecognize.clicked.connect(self.recognize)
        self.ui.btnShowDB.clicked.connect(self.show_db)
        self.ui.btnDelete.clicked.connect(self.delete_user)
        self.ui.btnExit.clicked.connect(self.close)

    def log(self, text):
        """–í—ã–≤–æ–¥ —Ç–µ–∫—Å—Ç–∞ –≤ –æ–∫–Ω–æ —Å—Ç–∞—Ç—É—Å–∞"""
        self.ui.textStatus.append(text)

    def register_user(self):
        name, ok = QtWidgets.QInputDialog.getText(self, "–†–µ–≥–∏—Å—Ç—Ä–∞—Ü–∏—è", "–í–≤–µ–¥–∏—Ç–µ –∏–º—è:")
        if ok and name:
            self.recognizer.register_face(name)
            self.log(f"‚úÖ –ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å {name} –∑–∞—Ä–µ–≥–∏—Å—Ç—Ä–∏—Ä–æ–≤–∞–Ω")

    def recognize(self):
        self.log("üöÄ –ó–∞–ø—É—Å–∫ —Ä–∞—Å–ø–æ–∑–Ω–∞–≤–∞–Ω–∏—è...")
        self.recognizer.recognize()
        self.log("üõë –†–∞—Å–ø–æ–∑–Ω–∞–≤–∞–Ω–∏–µ –∑–∞–≤–µ—Ä—à–µ–Ω–æ")

    def show_db(self):
        if self.recognizer.db.has_faces():
            users = "\n".join(self.recognizer.db.all_faces().keys())
            self.log("üë• –í –±–∞–∑–µ:\n" + users)
        else:
            self.log("‚ùå –ë–∞–∑–∞ –ø—É—Å—Ç–∞")

    def delete_user(self):
        if not self.recognizer.db.has_faces():
            self.log("‚ùå –ë–∞–∑–∞ –ø—É—Å—Ç–∞")
            return

        names = list(self.recognizer.db.all_faces().keys())
        name, ok = QtWidgets.QInputDialog.getItem(self, "–£–¥–∞–ª–µ–Ω–∏–µ", "–í—ã–±–µ—Ä–∏—Ç–µ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è:", names, 0, False)
        if ok:
            self.recognizer.db.remove_face(name)
            self.log(f"üóë –ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å {name} —É–¥–∞–ª—ë–Ω")


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = MainApp()
    window.show()
    sys.exit(app.exec_())
